Add implementations for functions and tests

non-binary opened the pull request:
WIP.

non-binary commented:
@zkat Can you please take a look at https://github.com/non-binary/http-cache-semantics/blob/18a4e26bf201257c60f2f1cb587b0dc4f3882c4a/src/lib.rs#L192 and let me know what I'm doing wrong? 

I'm getting the error `^^ expecting a type here because of type ascription`

```rust
if 407 == response_code {
    mock_response["headers"] = {"proxy-authenticate": "Basic realm=\"protected area\"".to_string()}
} else if 401 == response_code {
    mock_response["headers"] = {"www-authenticate": "Basic realm=\"protected area\"".to_string()}
} else if 204 == response_code || 205 == response_code {
    mock_response = {"body": "".to_string()}
}
```

I also tried without `.to_string()` after each so I'm not really sure. I googled around and all of the posts talk about a semi colon being where it shouldn't but I don't see that being the case here.

Full `cargo test` log here: https://gist.github.com/non-binary/6a0f01ca197b6068eb6fd246ed2f1453.

Any ideas? Thanks!

larsgw commented:
Are the JSON values in the if statements missing `json!` macro calls? Adding those seems to have fixed the errors on my end.

non-binary commented:
Thanks @larsgw! It now works for me as well. 💃 

With that said, can you please take a look at the function and clarify/let me know if the JSON value lines in the if statements will replace the initial data or just add to it/update it if the key already exists?

    fn assert_cached(should_put: bool, response_code: i32) {
        let expected_response_code = response_code;

        let mut mock_response = json!({
            "headers": {
                "last-modified": format_date(-105, 1),
                "expires": format_date(1, 3600),
                "www.authenticate": "challenge"
            },
            "status": response_code,
            "body": "ABCDE",
        });

        if 407 == response_code {
            mock_response["headers"] = json!({"proxy-authenticate": "Basic realm=\"protected area\""});
        } else if 401 == response_code {
            mock_response["headers"] = json!({"www-authenticate": "Basic realm=\"protected area\""});
        } else if 204 == response_code || 205 == response_code {
            mock_response = json!({"body": ""});
        }

        let mut request = json!({
            "url": "/",
            "headers": {}
        });

        // instantiate CachePolicy struct properly once it is implemented
        // let cache = CachePolicy();
        // cache.storable() -> bool

        assert_eq!(should_put, false);
    }

Thank you!

larsgw commented:
No problem!

> With that said, can you please take a look at the function and clarify/let me know if the JSON value lines in the if statements will replace the initial data or just add to it/update it if the key already exists?

That's what I'm trying to figure out, this is my first time working with Rust. The serialized results:

```js
// assert_cached(false, 100)
{"body":"ABCDE","headers":{"expires":"unimplemented","last-modified":"unimplemented","www.authenticate":"challenge"},"status":100}
// assert_cached(false, 407)
{"body":"ABCDE","headers":{"proxy-authenticate":"Basic realm=\"protected area\""},"status":407}
```

So it looks like it's overwriting.

larsgw commented:
This seems to work.

```diff
-            mock_response["headers"] = {"proxy-authenticate": "Basic realm=\"protected area\"".to_string()}
+            mock_response["headers"]["proxy-authenticate"] = json!("Basic realm=\"protected area\"")
```

non-binary commented:
'doh, that makes sense haha. Awesome. Thank you for the help! I really appreciate it. My background is Swift so struggling a bit with syntax.

non-binary commented:
@larsgw @zkat 

I'm having a hard time with Regex. This is something I struggle with regularly but at least I'm somewhat familiar with syntax and frameworks in Swift. Can you help me figure this out?

Here's the function I am trying to port: https://github.com/kornelski/http-cache-semantics/blob/master/index.js#L54.

What is this trying to do? I looked it up on https://regex101.com/ but I'm still quite confused.

```javascript
const parts = header.trim().split(/\s*,\s*/); // TODO: lame parsing
for (const part of parts) {
  const [k, v] = part.split(/\s*=\s*/, 2);
  cc[k] = v === undefined ? true : v.replace(/^"|"$/g, ""); // TODO: lame unquoting
}
```

Is it the first one just trying to break up items separated by commas into an array? Similarly with the second one but with equals signs in that case?

larsgw commented:
Yep, the first one breaks up items separated by commas, and any amount of surrounding whitespace if needed. The second one does the same for equals signs, but only once: the second parameter limits the total number of strings in the resulting array (never heard of that one before).

```js
'a=b, c= d ,e =f=g'

// ↓ split(/\s*,\s*/)

['a=b', 'c= d', 'e =f=g']

// ↓ split(/\s*=\s*/, 2)

[['a', 'b'], ['c', 'd'], ['e', 'f=g']]
```

non-binary commented:
@larsgw Okay, thank you for explaining this to me. Regex is hard 😫 How would the above JavaScript code be written in Rust? Do I have to use the Regex crate or is there a built in way to do it?

larsgw commented:
Something like this?

```rust
use std::collections::HashMap;

fn parse_cache_control (header: &str) -> HashMap<&str, &str> {
        let mut cache_control = HashMap::new();

        let parts: Vec<&str> = header.split(',').collect();
        for part in parts {
                let pair: Vec<&str> = part.splitn(2, '=').collect();
                let key = pair[0].trim();
                let value = pair[1].trim();
                cache_control.insert(key, value);
        }

        cache_control
}
```

Disclaimer: I haven't gotten to the ownership part of the tutorial yet, so I don't know the whether `&str` is okay in this case.

non-binary commented:
Thanks! I'll see what I can do with this. I appreciate the help. 💃 ❤️

non-binary commented:
I've implemented all of the OK HTTP tests. Will continue working on the other tests tomorrow/next few days. I want to finish implementing the tests before I move onto the real source code.

non-binary commented:
I've implemented all of the Request tests.

non-binary commented:
@larsgw @zkat I'm struggling with determining the best way to do this. I would appreciate any insight/recommendations on how I should handle this:

    pub struct CachePolicy {
        pub request: Value,
        pub response: Value,
        pub shared: Option<bool>,
        // and 12 more that I've excluded for this example...
    }

There are a lot of fields in the struct that have optional values or at least not every test implements them. So for example when I have:

    #[test]
    fn test_that_tests_the_things() {
        let cache = CachePolicy {
            request: json!({"headers": {}}),
            response: json!({"headers": {},
            }),
            shared: None
        };

        assert!(true);
    }

If I comment out `shared: None` I get `^^^^^^^^^^^ missing 'shared'` which means I would have to include every unused field in each test, which is...not great...

In the languages I'm used to you can just not include the values that aren't being used if they're associated to `nil`.

I _could_ create a function like `CachePolicy::new(req, res, etc)` but that's not great.

Is there any way to do this built into Structs/Rust?

Thanks in advance!

larsgw commented:
Something like https://users.rust-lang.org/t/idiomatic-way-to-construct-object-with-some-non-required-fields/8078/4 maybe?

non-binary commented:
Thanks for sharing this resource. I think I will proceed with this https://users.rust-lang.org/t/idiomatic-way-to-construct-object-with-some-non-required-fields/8078/5 way of doing it. Kind of unfortunate that Rust doesn’t have as powerful Struct features as Swift - would’ve come in handy.

non-binary commented:
I've been sick and busy/stressed about other things. I'm going to try to get back on this soon. Sorry for it taking so long. 😬 😞

zkat commented:
Don't sweat it. Please keep taking care of yourself 💚

non-binary commented:
Doing some work on this today.
